petastorm/arrow_reader_worker.py

```diff
@@ -174,6 +180,29 @@ def _load_rows(self, pq_file, piece, shuffle_row_drop_range):
             for field_to_remove in set(transformed_result.columns) & set(self._transform_spec.removed_fields):
                 del transformed_result[field_to_remove]
 
+            transformed_result_column_set = set(transformed_result.columns)
+            transformed_schema_column_set = set([f.name for f in self._transformed_schema.fields.values()])
+
+            if transformed_result_column_set != transformed_schema_column_set:
+                raise ValueError('Transformed result columns ({rc}) do not match required schema columns({sc})'
+                                 .format(rc=','.join(transformed_result_column_set),
+                                         sc=','.join(transformed_schema_column_set)))
+
+            def check_shape_and_ravel(x, field):
```

Comment:
Probably a matter of personal taste, but maybe we take this out to be a private function? My scare is unintentional capturing of some outer-function state. It's also more testable when it's not an embedded function.

BTW, could be a good idea to unittest different branches in the function.